File: Common/Graph/src/main/java/com/greentree/common/graph/algorithm/cycle/SmartCycleFinder.java
package com.greentree.common.graph.algorithm.cycle;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.greentree.common.graph.Graph;

public class SmartCycleFinder<V> {
	
	private final List<JointCycle<V>> res = new ArrayList<>();
	
	public SmartCycleFinder(Graph<V> graph) {
		Map<V, Boolean> color = new HashMap<>();
		for(V v : graph)
			color.put(v, false);
		for(V v : graph) {
			dfs(graph, new HashMap<>(), color, v);
			break;
		}
		res.removeIf(e->(e.size() < 3));
	}
	
	public List<JointCycle<V>> get() {
		return res;
	}
	
	private void dfs(Graph<V> graph, Map<V, V> parent, Map<V, Boolean> color, V v) {
		color.put(v, true);
		for(V to : graph.getJoints(v))
			if(!color.get(to)) {
				parent.put(to, v);
				dfs(graph, parent, color, to);
			}else {
				List<V> cycle = new ArrayList<>();
				for(V t = v;; t = parent.get(t)) {
					cycle.add(t);
					if(t.equals(to))
						break;
				}
				JointCycle<V> c0 = new JointCycle<>(cycle);
				if(!res.contains(c0))
					res.add(c0);
			}
		color.put(v, false);
	}
}

File: Commons/Graph/src/main/java/com/greentree/commons/graph/algorithm/cycle/SmartCycleFinder.kt
package com.greentree.commons.graph.algorithm.cycle

import com.greentree.commons.graph.FiniteGraph

class SmartCycleFinder<V : Any>(graph: FiniteGraph<V>) {

	private val res: MutableList<JointCycle<V>> = ArrayList()

	init {
		val color: MutableMap<V, Boolean> = HashMap()
		for(v in graph) color[v] = false
		for(v in graph) {
			dfs(graph, HashMap(), color, v)
			break
		}
		res.removeIf { e: JointCycle<V> -> (e.size < 3) }
	}

	fun get(): List<JointCycle<V>> {
		return res
	}

	private fun dfs(graph: FiniteGraph<V>, parent: MutableMap<V?, V>, color: MutableMap<V, Boolean>, v: V) {
		color[v] = true
		for(to in graph.getAdjacencyIterable(v)) if(!color[to]!!) {
			parent[to] = v
			dfs(graph, parent, color, to)
		} else {
			val cycle: MutableList<V> = ArrayList()
			var t: V = v
			while(true) {
				cycle.add(t)
				if(t == to) break
				t = parent[t]!!
			}
			val c0 = JointCycle(cycle)
			if(!res.contains(c0)) res.add(c0)
		}
		color[v] = false
	}
}
